python/cuml/neighbors/nearest_neighbors_mg.pyx

```diff
@@ -0,0 +1,350 @@
+            = <PartDescriptor*><size_t>query_desc
+        for elm in range(query_vec_c.size()):
+            free(query_vec_c.at(elm))
+        free(query_vec_c)
+        free(query_desc_c)
+
+        cdef vector[int64Data_t *] *out_i_vec_c \
+            = <vector[int64Data_t *]*><size_t>out_i_vec
+        cdef int64Data_t *del_idx_ptr
+        for elm in range(out_i_vec_c.size()):
+            del_idx_ptr = out_i_vec_c.at(elm)
+            del del_idx_ptr
+        free(out_i_vec_c)
+
+        cdef vector[floatData_t *] *out_d_vec_c \
+            = <vector[floatData_t *]*><size_t>out_d_vec
+        cdef floatData_t *del_ptr
+        for elm in range(out_d_vec_c.size()):
+            del_ptr = out_d_vec_c.at(elm)
+            del del_ptr
+        free(out_d_vec_c)
+
+        cdef vector[RankSizePair *] *local_index_parts_c \
+            = <vector[RankSizePair *]*><size_t>local_index_parts
+        for elm in range(local_index_parts_c.size()):
+            free(local_index_parts_c.at(elm))
+        free(local_index_parts_c)
+
+        cdef vector[RankSizePair *] *local_query_parts_c \
+            = <vector[RankSizePair *]*><size_t>local_query_parts
+        for elm in range(local_query_parts_c.size()):
+            free(local_query_parts_c.at(elm))
+        free(local_query_parts_c)
+
+    def _freeFloatD(self, data):
+        cdef uintptr_t data_ptr = data
+        cdef vector[floatData_t*] *d = <vector[floatData_t*]*>data_ptr
+        for x_i in range(d.size()):
+            free(d.at(x_i))
+        free(d)
+
+    def kneighbors(self, indices, index_m, n, index_partsToRanks,
+                   queries, query_m, query_partsToRanks,
+                   rank, n_neighbors=None, convert_dtype=True):
+        """
+        Query the kneighbors of an index
+        :param indices: [__cuda_array_interface__] of local index partitions
+        :param index_m: number of total index rows
+        :param n: number of columns
+        :param index_partsToRanks: mappings of index partitions to ranks
+        :param queries: [__cuda_array_interface__] of local query partitions
+        :param query_m: number of total query rows
+        :param query_partsToRanks: mappings of query partitions to ranks
+        :param rank: int rank of current worker
+        :param n_neighbors: int number of nearest neighbors to query
+        :return:
+        """
+
+        n_neighbors = self.n_neighbors if n_neighbors is None else n_neighbors
+
+        self.n_dims = n
+
+        cdef cumlHandle* handle_ = <cumlHandle*><size_t>self.handle.getHandle()
+
+        cdef vector[RankSizePair*] *index_vec = new vector[RankSizePair*]()
+        cdef vector[RankSizePair*] *query_vec = new vector[RankSizePair*]()
+
+        # Build device arrays for local partitions
+        query_ints = []
+        index_ints = []
+        for arr in queries:
+            X_m, input_ptr, n_rows, n_cols, dtype = \
+                input_to_dev_array(arr, order="F",
```

Comment:
Do we sometimes use order=F and sometimes order=C? Seems like order=C is just used for labels+outputs, but I'm not sure I understand why

Reply:
One of the nice benefits to updating FAISS is that the brute force kNN now supports column major inputs, which we should be striving to support as much as possible in this layer to alleviate the need for copies. The output, unfortunately, is still row major.